### ISSUE TITLE
Make A<T>.Ignored throw an exception when it is invoked

### ISSUE THREAD

**Issue by thomaslevesque** [FakeItEasy/FakeItEasy#559]
I know, that sounds crazy... but it's not ;)

`A<T>.Ignored` (or its shortcut `A<T>._`) is used in call specifications to denote that any value of this argument should be a match. If you have long type names (e.g. `IEnumerable<CustomerInvoiceDetailLine>` or something like that), it can be tempting to do this:

```
var any = A<IEnumerable<CustomerInvoiceDetailLine>>.Ignored;
A.CallTo(() => x.PrintInvoiceDetails(any))....
```

However, this doesn't work, because FIE analyzes the expression in `A.CallTo` to detect calls to `A<T>.Ignored`, and in this case the expression doesn't contain it. `A<T>.Ignored` is not meant to actually be called, only to be analyzed as part of the call specification.

This can be very confusing to many users, so I think `A<T>.Ignored` should throw an exception (with a message explaining the problem) when it is invoked. It would be better to be able to detect it at compile time, but I don't think it's possible.

(this was inspired by [this question about Moq](http://stackoverflow.com/q/33669722/98713), which has the same issue)

**Comment by blairconrad** [comment]
Not crazy at all. I've not analyzed,but this sounds like a great idea  to me, @thomaslevesque.
The now-private `Now` [works the same way](https://github.com/FakeItEasy/FakeItEasy/blob/bd57a05c0de5ff37609230fd1b8bc4ec45bc9d4d/Source/FakeItEasy/Core/Raise.of.TEventArgs.cs#L72).

**Comment by blairconrad** [comment]
@thomaslevesque, I may have been hasty. And maybe I'm being hasty now, but you said

> FIE analyzes that expression to recognize `A<T>.Ignored` or `A<T>._`

But really `Ignored` looks like this:

```c#
public static T Ignored
{
    get { return That.Matches(x => true, x => x.Write("Ignored")); }
}
```

So the code is executed, but we rely on it being executed within the context of the `A.CallTo`, as that's when we capture the [event raised by `Matches`](https://github.com/FakeItEasy/FakeItEasy/blob/bd57a05c0de5ff37609230fd1b8bc4ec45bc9d4d/Source/FakeItEasy/Core/DefaultArgumentConstraintManager.cs#L20):

```c#
public T Matches(Func<T, bool> predicate, Action<IOutputWriter> descriptionWriter)
{
    this.onConstraintCreated(new MatchesConstraint(predicate, descriptionWriter));
    return default(T);
}
```

**Comment by thomaslevesque** [comment]
OK, it seems it might be a bit more complex than I thought...

**Comment by thomaslevesque** [comment]
I thought it could be detected in `ArgumentConstraintTrap.ReportTrappedConstraint`, but apparently there are valid scenarios where `trappedConstraints` is null (two tests fail if I throw an exception when it's null)

**Comment by blairconrad** [comment]
Oh, novel approach. Now I'm curious about which tests.
Don't do this to me while I'm at work!

**Comment by blairconrad** [comment]
@thomaslevesque, I took a quick peek (thanks to slow compiles), and I suspect the failing tests may not be ones that need to keep working as is. That is, they may be performing the equivalent action of "saving away an `A<T>.Ignored` for later use" for no real reason.

I also wonder if we couldn't make it possible to save away the constraint for later use by adding an implicit conversion operator to the thing that `Matches` returns, but I've not tried it…

**Comment by thomaslevesque** [comment]
I'm looking at `ExpressionArgumentConstraintFactoryTests.Should_get_aggregate_constraint_when_multiple_items_are_passed_to_parameters_array`, and I don't really understand what's going on... At one point we end up calling `ReportTrappedConstraint`, but not from within `TrapConstraints`, so `trappedConstraints` is null. Not sure what this means, or what this test is supposed to do.

If we can just drop those tests, that's great, but I don't know FakeItEasy's core well enough to understand the implications.

**Comment by blairconrad** [comment]
Yeah. I was looking at that too. Ultimately, it appears to be checking that when multiple constraint items are passed in the expression, we parse them properly and get one "aggregate constraint" - just a constraint that embodies all the little ones.
The way that we make these calls in the unit tests may not reflect legitimate usage.

Since I think we only want to use Ignored when specifying constraints or checking for completed calls that matched a constraint, my inclination would be to try to write specs to cover those two cases (using the `A.CallTo` syntax as well as `Fake<T>.CallsTo` for future calls), checking to make sure that it covered this aggregate constraint case (perhaps by a different path) and then replacing the confusing test with the specs, which will be in "user language" so we'll know what they're actually testing.
If nothing else, I think this test (maybe all the tests in the file) might live better as tests for `ExpressionArgumentConstraintFactory`. I think that could alleviate the crazy faking and then we'd have `ReportTrappedConstraint` called from `TrapConstraints`. I'm going to have to knock off pretty soon, but I may give that a quick try…

**Comment by blairconrad** [comment]
Aha. @thomaslevesque , I believe `CallMatchingSpecs.ParameterArrays` covers the functionality that `ExpressionArgumentConstraintFactoryTests.Should_get_aggregate_constraint_when_multiple_items_are_passed_to_parameters_array` cares about.
Toodles.

**Comment by blairconrad** [comment]
> I also wonder if we couldn't make it possible to save away the constraint for later use by adding an implicit conversion operator to the thing that Matches returns, but I've not tried it…

I tried it. And of course it didn't work, because we can't add an implicit cast operator that casts to an interface.

**Comment by thomaslevesque** [comment]
So, what do you think? Should we just drop those tests to allow the new behavior?

**Comment by blairconrad** [comment]
I think that's appropriate for `Should_get_aggregate_constraint_when_multiple_items_are_passed_to_parameters_array`. Its functionality is clearly covered elsewhere.

I'm slightly nervous about `Should_not_fail_when_reporting_trapped_constraint_outside_call_to_trap_constraints`. It was explicitly written to ensure that the change you propose would not work. That being said, I can't see why this behaviour is desirable. It was added in c75e0c6d75c287749ecd983c33e5097633539d58, but before the trapper was hooked up to any code accessible from the outside.

However, I really can't see a point to using `A<T>.That.Matches` when we're not trapping arguments, so I think that the test can be dropped (in favour of a unit test or spec that ensures we throw).
But perhaps @patrik-hagne or the other owners know of a reason why this is not the case.

**Comment by patrik-hagne** [comment]
@blairconrad I can't remember why that test is explicit about not failing. Could it have anything to do with NextCall.To...?

**Comment by blairconrad** [comment]
Oh, excellent point, @patrik-hagne. We'll check that out. Thanks!

**Comment by blairconrad** [comment]
@patrik-hagne, I think it does not have anything to do with `NextCall.To`, or at least it doesn't as the code is in master right now.
`NextCall.To` either takes a custom matcher via `WhenArgumentsMatch(Func<ArgumentCollection, bool> argumentsPredicate)`, or it relies on equality matching the arguments supplied in the next call.
In either case, there's no active argument trap, and the event raised from `That.Matches` does not fire (with the proposed change in this issue, it would, causing an error).
The current behaviour with `That.Matches` in the `NextCall` configuration is to match the `default(T)` that's created, so we get this behaviour:

```c#
[Test]
public void CanUseMatchesWithNextCall()
{
    var foo = A.Fake<IFoo>();
    NextCall.To(foo).Throws(new Exception("Blair"));
    foo.Bar(A<object>.That.Matches(o=>((int)o) == 5));
    
    foo.Bar(5); // does not throw
    foo.Bar(null); // null == default(object), so we throw
}
```

I think most users would be confused by `foo.Bar(5)` not throwing and so we'd still benefit from the change that @thomaslevesque proposes.

**Comment by thomaslevesque** [comment]
So, `NextCall.To` doesn't support `That` or `Ignored` constraints, but trying to use them doesn't throw? This should probably be changed as well. So I guess we can just remove both breaking tests.

**Comment by blairconrad** [comment]
That makes sense to me.

**Comment by thomaslevesque** [comment]
Eventually I didn't remove any of the breaking tests.

I changed `Should_not_fail_when_reporting_trapped_constraint_outside_call_to_trap_constraints` to `Should_fail_when_reporting_trapped_constraint_outside_call_to_trap_constraints`; now instead of asserting that `ReportTrappedConstraint` does _not_ throw, it asserts that it _does_ throw, since that's what we want.

Regarding `Should_get_aggregate_constraint_when_multiple_items_are_passed_to_parameters_array`, it was failing because it was using a fake `IArgumentConstraintTrapper`, but the constraints were trapped with the static `ReportTrappedConstraint` method of the real `ArgumentConstraintTrap` class... So the `trappedConstraints` field wasn't set, and it failed. So I just used a real `ArgumentConstraintTrap` to invoke the constraint expression.

**Comment by adamralph** [comment]
Should we call this breaking? I guess if anyone is invoking it, then that code simply redundant, but I guess it will cause a test failure after upgrade...

**Comment by blairconrad** [comment]
I'm usually a big fan of making things breaking, but in this case, I'd say not.

I _suppose_ someone is using say `Ignored`, and the value is stored as `default(T)` and then later the value is used in an `A.CallTo` and it _happens_ to work out, because their production code is passing `default(T)` in as the value of the argument in question, but still, that's a bug in their test. If the production code changes to pass in something else, their test will fail, when it shouldn't, because they're not getting the predicate they think they are.

Calling these methods outside of `A.CallTo` is a clear misuse. We're just helping people discover these misuses early.

**Comment by adamralph** [comment]
OK, I'm convinced :smile: (I wasn't at all certain about the suggestion.)

**Comment by thomaslevesque** [comment]
I agree with @blairconrad that it's not really breaking, but we should probably mention it in the release notes, to anticipate questions about it

**Comment by blairconrad** [comment]
I'd updated the release notes this morning. You think a more dire warning?

**Comment by thomaslevesque** [comment]
> I'd updated the release notes this morning. You think a more dire warning is needed?

No, just mentioning it enough I think. But it should be more specific I think; `A.Ignored` only throws when invoked outside of `A.CallTo`

**Comment by blairconrad** [comment]
Sounds good to me. Go ahead! :wink:

**Comment by blairconrad** [comment]
This issue has been fixed in [release 2.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/2.0.0).

**Comment by Jameskmonger** [comment]
Do you know which commit fixed this, @blairconrad? I had a look but can't see anything :)

**Comment by blairconrad** [comment]
The commit appears to be https://github.com/FakeItEasy/FakeItEasy/pull/619/commits/cc0fe00a144f36091a1e59b979fbe4112695020c.

**Comment by blairconrad** [comment]
@Jameskmonger, I was wrong. I thought my superseding PR (#619) was merged, but @thomaslevesque's original PR (#611) was.
The commit is 48dc7682e33a7e666b2ed7bebaf7325863313b75.
